File: code/packages/java/ct-compare/src/test/java/com/codingadventures/ctcompare/CtCompareTest.java
package com.codingadventures.ctcompare;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class CtCompareTest {
    @Test
    void ctEqMatchesByteEquality() {
        assertTrue(CtCompare.ctEq("abcdef".getBytes(), "abcdef".getBytes()));
        assertTrue(CtCompare.ctEq(new byte[0], new byte[0]));
        assertFalse(CtCompare.ctEq("abcdef".getBytes(), "abcdeg".getBytes()));
        assertFalse(CtCompare.ctEq("abcdef".getBytes(), "bbcdef".getBytes()));
        assertFalse(CtCompare.ctEq("abc".getBytes(), "abcd".getBytes()));
    }

    @Test
    void ctEqDetectsEverySingleBitPosition() {
        byte[] baseline = new byte[32];
        java.util.Arrays.fill(baseline, (byte) 0x42);
        for (int index = 0; index < baseline.length; index++) {
            for (int bit = 0; bit < 8; bit++) {
                byte[] flipped = baseline.clone();
                flipped[index] ^= (byte) (1 << bit);
                assertFalse(CtCompare.ctEq(baseline, flipped));
            }
        }
    }

    @Test
    void ctEqFixedIsDynamicAlias() {
        assertTrue(CtCompare.ctEqFixed(new byte[16], new byte[16]));
        byte[] different = new byte[16];
        different[15] = 1;
        assertFalse(CtCompare.ctEqFixed(new byte[16], different));
    }

    @Test
    void ctSelectBytesChoosesWithoutMutatingInputs() {
        byte[] left = new byte[256];
        byte[] right = new byte[256];
        for (int i = 0; i < 256; i++) {
            left[i] = (byte) i;
            right[i] = (byte) (255 - i);
        }

        assertArrayEquals(left, CtCompare.ctSelectBytes(left, right, true));
        assertArrayEquals(right, CtCompare.ctSelectBytes(left, right, false));
        assertArrayEquals(new byte[0], CtCompare.ctSelectBytes(new byte[0], new byte[0], true));
        assertThrows(IllegalArgumentException.class, () -> CtCompare.ctSelectBytes(new byte[1], new byte[2], true));
    }

    @Test
    void ctEqU64HandlesEdges() {
        assertTrue(CtCompare.ctEqU64(0L, 0L));
        assertTrue(CtCompare.ctEqU64(-1L, -1L));
        assertFalse(CtCompare.ctEqU64(0L, Long.MIN_VALUE));

        long baseline = 0x1234_5678_9ABC_DEF0L;
        for (int bit = 0; bit < 64; bit++) {
            assertFalse(CtCompare.ctEqU64(baseline, baseline ^ (1L << bit)));
        }
    }
}

File: code/packages/kotlin/ct-compare/src/test/kotlin/com/codingadventures/ctcompare/CtCompareTest.kt
package com.codingadventures.ctcompare

import kotlin.test.Test
import kotlin.test.assertContentEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class CtCompareTest {
    @Test
    fun ctEqMatchesByteEquality() {
        assertTrue(CtCompare.ctEq("abcdef".encodeToByteArray(), "abcdef".encodeToByteArray()))
        assertTrue(CtCompare.ctEq(ByteArray(0), ByteArray(0)))
        assertFalse(CtCompare.ctEq("abcdef".encodeToByteArray(), "abcdeg".encodeToByteArray()))
        assertFalse(CtCompare.ctEq("abcdef".encodeToByteArray(), "bbcdef".encodeToByteArray()))
        assertFalse(CtCompare.ctEq("abc".encodeToByteArray(), "abcd".encodeToByteArray()))
    }

    @Test
    fun ctEqDetectsEverySingleBitPosition() {
        val baseline = ByteArray(32) { 0x42 }
        for (index in baseline.indices) {
            for (bit in 0 until 8) {
                val flipped = baseline.copyOf()
                flipped[index] = (flipped[index].toInt() xor (1 shl bit)).toByte()
                assertFalse(CtCompare.ctEq(baseline, flipped))
            }
        }
    }

    @Test
    fun ctEqFixedIsDynamicAlias() {
        assertTrue(CtCompare.ctEqFixed(ByteArray(16), ByteArray(16)))
        val different = ByteArray(16)
        different[15] = 1
        assertFalse(CtCompare.ctEqFixed(ByteArray(16), different))
    }

    @Test
    fun ctSelectBytesChoosesWithoutMutatingInputs() {
        val left = ByteArray(256) { it.toByte() }
        val right = ByteArray(256) { (255 - it).toByte() }

        assertContentEquals(left, CtCompare.ctSelectBytes(left, right, true))
        assertContentEquals(right, CtCompare.ctSelectBytes(left, right, false))
        assertContentEquals(ByteArray(0), CtCompare.ctSelectBytes(ByteArray(0), ByteArray(0), true))
        assertFailsWith<IllegalArgumentException> {
            CtCompare.ctSelectBytes(ByteArray(1), ByteArray(2), true)
        }
    }

    @Test
    fun ctEqU64HandlesEdges() {
        assertTrue(CtCompare.ctEqU64(0L, 0L))
        assertTrue(CtCompare.ctEqU64(-1L, -1L))
        assertFalse(CtCompare.ctEqU64(0L, Long.MIN_VALUE))

        val baseline = 0x1234_5678_9ABC_DEF0L
        for (bit in 0 until 64) {
            assertFalse(CtCompare.ctEqU64(baseline, baseline xor (1L shl bit)))
        }
    }
}
